src/theory/quantifiers/sygus_grammar_norm.h

```diff
@@ -14,137 +14,439 @@
+   * of monomials, as seen above. The operator for "0" is rebuilt as is (the
+   * default behaviour of operators not selected for transformations).
+   *
+   * recursion depth is limited by the height of the types, which is small
+   */
+  TypeNode normalizeSygusRec(TypeNode tn,
+                             const Datatype& dt,
+                             std::vector<unsigned> op_pos);
 
-  /** Traverses the datatype representation of src_tn and collects the types it
-   * contains
+  /** wrapper for the above function
    *
-   * For each new typenode a TypeObject is created, with an unresolved type and
-   * a datatype to be later resolved and constructed, respectively. These are
-   * accumulated in tos.
+   * invoked when all operators of "tn" are to be considered for normalization
+   */
+  TypeNode normalizeSygusRec(TypeNode tn);
+
+  /** List of variable inputs of function-to-synthesize.
    *
-   * tn_to_unres maps the sygus types that the typenodes stand for into
-   * the unresolved type to be built for the typenodes normalizations.
+   * This information is needed in the construction of each datatype
+   * representation of type nodes contained in the type node being normalized
+   */
+  TNode d_sygus_vars;
+
+  /* adds datatypes /unresolved types to global accumulators */
+  inline void addDatatype(Datatype dt) { d_dt_all.push_back(dt); }
+  inline void addUnresType(Type t) { d_unres_t_all.insert(t); }
+
+  /* Retrives, or, if none, creates, store and return, the node for the identity
+   * operator (\lambda x. x) for the given type node */
+  static inline Node getIdOp(TypeNode tn)
+  {
+    auto it = d_tn_to_id.find(tn);
+    if (it == d_tn_to_id.end())
+    {
+      return addIdOp(tn);
+    }
+    return it->second;
+  }
+
+ private:
+  /** reference to quantifier engine */
+  QuantifiersEngine* d_qe;
+  /** sygus term database associated with this utility */
+  TermDbSygus* d_tds;
+  /* Datatypes to be resolved */
+  std::vector<Datatype> d_dt_all;
+  /* Types to be resolved */
+  std::set<Type> d_unres_t_all;
+  /* Associates type nodes with OpPosTries */
+  std::map<TypeNode, OpPosTrie> d_tries;
+
+  /** infers a transformation for normalizing dt when allowed to use the
+   * operators in the positions op_pos.
    *
-   * visited caches visited nodes
+   * TODO: #1304: Infer more complex transformations
    */
-  void collectInfoFor(TypeNode src_tn,
-                      std::vector<TypeObject>& tos,
-                      std::map<TypeNode, Type>& tn_to_unres,
-                      std::map<TypeNode, bool>& visited);
-};
+  Transf* inferTransf(TypeNode tn,
+                      const Datatype& dt,
+                      std::vector<unsigned>& op_pos);
+
+  /* Map of type nodes into their identity operators (\lambda x. x) */
+  static std::map<TypeNode, Node> d_tn_to_id;
+
+  /** associates identity operator with the given type node
+   *
+   * returns the built node
+   */
+  static inline Node addIdOp(TypeNode tn)
```

Comment:
Rewrite this so it is "getIdOp", which first checks if d_tn_to_id has an entry for tn.  This way we only need to remember to call this function and the name of the map is keep local and only referred to in this function.